External/pubchem_shape/PubChemShape.cpp

```diff
@@ -0,0 +1,424 @@
+        "[r]1[r][r][r][r]1",
+        "[r]1[r][r][r][r][r]1",
+        "[r]1[r][r][r][r][r][r]1",
+    },  // rings
+        //    "[a]",                                                  //
+        //    Aromatic
+        //    "[F,Cl,Br,I]",                                          // Halogen
+    {"[#7;+,\
+$([N;H2&+0][$([C,a]);!$([C,a](=O))]),\
+$([N;H1&+0]([$([C,a]);!$([C,a](=O))])[$([C,a]);!$([C,a](=O))]),\
+$([N;H0&+0]([C;!$(C(=O))])([C;!$(C(=O))])[C;!$(C(=O))])]"},  // Basic
+    {"[$([C,S](=[O,S,P])-[O;H1,-1])]"}                       // Acidic
+};
+std::vector<std::vector<const ROMol *>> *getPh4Patterns() {
+  static std::unique_ptr<std::vector<std::vector<const ROMol *>>> patterns;
+  if (!patterns) {
+    patterns.reset(new std::vector<std::vector<const ROMol *>>());
+    for (const auto &smartsV : smartsPatterns) {
+      std::vector<const ROMol *> v;
+      for (const auto &smarts : smartsV) {
+        const ROMol *matcher = pattern_flyweight(smarts).get().getMatcher();
+        CHECK_INVARIANT(matcher, "bad smarts");
+        v.push_back(matcher);
+      }
+      patterns->push_back(std::move(v));
+    }
+  }
+
+  return patterns.get();
+}
+
+// the conformer is translated to the origin
+ShapeInput PrepareConformer(const ROMol &mol, int confId, bool useColors) {
+  ShapeInput res;
+
+  // unpack features (PubChem-specific property from SDF)
+  // NOTE: this unpacking assumes that RWMol-atom-index = SDF-atom-number - 1
+  //   e.g. RWMol uses [0..N-1] and SDF uses [1..N], with atoms in the same
+  //   order
+
+  vector<pair<vector<unsigned int>, unsigned int>> feature_idx_type;
+
+  if (useColors) {
+    std::string features;
+    if (mol.getPropIfPresent("PUBCHEM_PHARMACOPHORE_FEATURES", features)) {
+      // regular atoms have type 0; feature "atoms" (features represented by a
+      // single point+radius) must have type > 0
+      static const map<string, unsigned int> atomTypes = {
+          {"acceptor", 1}, {"anion", 2},      {"cation", 3},
+          {"donor", 4},    {"hydrophobe", 5}, {"rings", 6},
+      };
+
+      istringstream iss(features);
+      string line;
+      unsigned int n = 0;
+      while (getline(iss, line)) {
+        if (n == 0) {
+          feature_idx_type.resize(stoul(line));
+        }
+
+        else {
+          unsigned int f = n - 1;
+          if (f >= feature_idx_type.size()) ERRORTHROW("Too many features");
+
+          istringstream iss2(line);
+          string token;
+          unsigned int t = 0;
+          while (getline(iss2, token, ' ')) {
+            if (t == 0) {
+              feature_idx_type[f].first.resize(stoul(token));
+            } else if (t <= feature_idx_type[f].first.size()) {
+              feature_idx_type[f].first[t - 1] = stoul(token) - 1;
+            } else {
+              map<string, unsigned int>::const_iterator type =
+                  atomTypes.find(token);
+              if (type == atomTypes.end()) ERRORTHROW("Invalid feature type");
```

Comment:
Add the token that was the invalid feature type to the error, this will help users debug,